pcdsdevices/ipm.py

```diff
@@ -49,15 +100,274 @@ def removed(self):
+    ch1 = Cpt(Wave8Channel, '', channel_index=1, kind='normal')
+    ch2 = Cpt(Wave8Channel, '', channel_index=2, kind='normal')
+    ch3 = Cpt(Wave8Channel, '', channel_index=3, kind='normal')
+    ch4 = Cpt(Wave8Channel, '', channel_index=4, kind='normal')
+    ch5 = Cpt(Wave8Channel, '', channel_index=5, kind='normal')
+    ch6 = Cpt(Wave8Channel, '', channel_index=6, kind='normal')
+    ch7 = Cpt(Wave8Channel, '', channel_index=7, kind='normal')
+    ch8 = Cpt(Wave8Channel, '', channel_index=8, kind='normal')
+    ch9 = Cpt(Wave8Channel, '', channel_index=9, kind='normal')
+    ch10 = Cpt(Wave8Channel, '', channel_index=10, kind='normal')
+    ch11 = Cpt(Wave8Channel, '', channel_index=11, kind='normal')
+    ch12 = Cpt(Wave8Channel, '', channel_index=12, kind='normal')
+    ch13 = Cpt(Wave8Channel, '', channel_index=13, kind='normal')
+    ch14 = Cpt(Wave8Channel, '', channel_index=14, kind='normal')
+    ch15 = Cpt(Wave8Channel, '', channel_index=15, kind='normal')
+
+    def __init__(self, prefix, *, name, prefix_ioc=None, **kwargs):
+        if not prefix_ioc:
+            self._prefix_ioc = 'IOC:%s' % prefix
+        else:
+            self._prefix_ioc = prefix_ioc
+        super().__init__(prefix, name=name, **kwargs)
+
+    def screen(self):
+        """
+        Function to call the (pyQT) screen for a wave8 box
+        """
+        executable = '/reg/g/pcds/pyps/apps/wave8/latest/wave8'
+        if shutil.which(executable) is None:
+            print('%s is not on path, we cannot start the screen' % executable)
+            return
+        os.system('%s --base %s --ioc %s --evr %s &' %
+                  (executable, self.prefix, self._prefix_ioc,
+                   self.prefix+':TRIG'))
+
+
+class IPM_Det():
+    """
+    Base class for IPM_IPIMB and IPM_Wave8. Not meant to be instantiated.
+    """
+
+    def isum(self):
+        return self.det.isum.get()
+
+    def xpos(self):
+        return self.det.xpos.get()
+
+    def ypos(self):
+        return self.det.ypos.get()
+
+    def channel(self, i=0):
+        if (i >= self._num_channels or i < 0):
+            raise ValueError("Invalid channel number!")
+        else:
+            return getattr(self.det, 'ch%d' % i).amplitude.get()
+
+    def channels(self):
+        return [getattr(self.det, 'ch%d' % i).amplitude.get()
+                for i in range(self._num_channels)]
+
+
+class IPM_IPIMB(IPMMotion, IPM_Det):
+    """
+%s
+
+    has a member ipimb which represents the ipimb box used for readout
+    """
+    __doc__ = __doc__ % (IPM_base) + basic_positioner_init
+
+    ipimb = FCpt(IPIMB, '{self.prefix_ipimb}', prefix_ioc='{self.prefix_ioc}')
+
+    # IPIMB's have four channels
+    _num_channels = 4
+
+    def __init__(self, prefix, *, name, **kwargs):
+        self.prefix_ipimb = kwargs.pop('prefix_ipimb')
+        self.prefix_ioc = kwargs.pop('prefix_ioc', None)
+        super().__init__(prefix, name=name, **kwargs)
+        self.det = self.ipimb
```

Comment:
the `self.det` setattr here plus the new parent is a :+1: implementation of removing the duplicated code